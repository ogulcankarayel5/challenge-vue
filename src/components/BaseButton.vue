<template>
  <component
    :is="tag"
    :href="to"
    @click="onClickButton"
    class="button"
    :class="btnClass"
  >
    <slot />
  </component>
</template>

<script>
export default {
  name: "BaseButton",
  props: {
    to: {
      type: String,
    },
    tag: {
      type: String,
      default: "button",
    },
    size: {
      type: String,
      default: "normal",
      validator: (value) => {
        return ["normal", "small"].indexOf(value) !== -1;
      },
    },
    variant: {
      type: String,
      default: "transparent",
      validator: (value) => {
        return ["primary", "transparent"].indexOf(value) !== -1;
      },
    },
  },
  methods: {
    onClickButton() {
      this.$emit("onClick");
    },
  },
  emits: ["onClick"],
  computed: {
    btnClass() {
      return {
        "button-normal": this.size === "normal",
        "button-small": this.size === "small",

        "button-transparent": this.variant === "transparent",
        "button-primary": this.variant === "primary",
      };
    },
  },
};
</script>

<style scoped lang="sass">
.button

  &-transparent
    background-color: transparent
  &-primary
    background-color: #000
    color: #fff
  &-normal
    width: 80px
    padding: 10px 0
</style>
