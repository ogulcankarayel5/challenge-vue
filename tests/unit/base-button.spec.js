import { mount } from "@vue/test-utils";
import BaseButton from "@/components/BaseButton.vue";

const factory = (props = {}, slots = {}) => {
  return mount(BaseButton, {
    props: { ...props },
    slots: { ...slots },
  });
};

describe("BaseButton.vue", () => {
  it("snapshot", () => {
    const wrapper = factory();
    expect(wrapper.element).toMatchSnapshot();
  });
  it("renders component", () => {
    const wrapper = factory();
    expect(wrapper.find("button")).toBeTruthy();
  });
  it("renders class correctly according to props", () => {
    const wrapper = factory({ size: "small", variant: "primary" });

    expect(wrapper.find("button").classes()).toContain("button-small");
    expect(wrapper.find("button").classes()).toContain("button-primary");
  });
  it("emits correctly", async () => {
    const wrapper = factory();
    wrapper.vm.$emit("onClick");
    await wrapper.vm.$nextTick();

    expect(wrapper.emitted().onClick).toBeTruthy();
  });
  it("renders slot", async () => {
    const wrapper = mount(BaseButton, {
      slots: {
        default: "<p>button</p>",
      },
    });

    expect(wrapper.html()).toContain("<p>button</p>");
  });
});
