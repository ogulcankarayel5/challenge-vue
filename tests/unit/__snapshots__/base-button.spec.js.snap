// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseButton.vue snapshot 1`] = `
<button
  class="button button-normal button-transparent"
>
  
  
</button>
`;
